### PR TITLE
LLVM module support for named structs

### LINKED ISSUE #237: LLVM module support for named structs
LLVM allows you to give names for structs.  It would be useful to be able to take advantage of that.

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
LLVM module support for named structs
Resolves #237 

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/print.llvm/llvm.bal [hunk 1]
```diff
@@ -179,10 +179,41 @@ public class Context {
...
         words.push(")");
         return constValueWithBody(destTy, words);
     }
+
+    public function structCreateNamed(string name, Type[] elementTypes) returns StructType {
+        string structName = "%" + escapeIdent(name);
+        if self.namedStructTypes.hasKey(structName) {
+            panic err:illegalArgument("This context already has a struct type by that name");
+        }
+        StructType ty = {name:structName, elementTypes:elementTypes.cloneReadOnly()};
```
Comment (jclark) [inline compiler/modules/print.llvm/llvm.bal]: Add spaces.

File: compiler/modules/print.llvm/common.bal [hunk 2]
```diff
@@ -39,11 +39,12 @@ public function arrayType(Type ty, int elementCount) returns ArrayType {
 
 // Corresponds to llvm::StructType
 public type StructType readonly & record {
+    string? name;
```
Comment (jclark) [inline compiler/modules/print.llvm/common.bal]: Why is `name` part of the struct type?

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/print.llvm/llvm.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/llvm.bal]: fixed

File: compiler/modules/print.llvm/common.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/common.bal]: I thought `name` to be a part of the type similar to it's element types.

File: compiler/modules/print.llvm/llvm.bal [hunk 3]
```diff
@@ -179,10 +179,41 @@ public class Context {
...
         words.push(")");
         return constValueWithBody(destTy, words);
     }
+
+    public function structCreateNamed(string name, Type[] elementTypes) returns StructType {
+        string structName = "%" + escapeIdent(name);
+        if self.namedStructTypes.hasKey(structName) {
+            panic err:illegalArgument("This context already has a struct type by that name");
+        }
+        StructType ty = { name:structName, elementTypes:elementTypes.cloneReadOnly() };
```
Comment (jclark) [inline compiler/modules/print.llvm/llvm.bal]: Space after colorn

File: compiler/modules/print.llvm/llvm.bal [hunk 4]
```diff
@@ -179,10 +179,41 @@ public class Context {
...
-        addrSpaceCastArgs(words, ptr, destTy);
+        addrSpaceCastArgs(words, ptr, destTy, self);
         words.push(")");
         return constValueWithBody(destTy, words);
     }
+
+    public function structCreateNamed(string name, Type[] elementTypes) returns StructType {
+        string structName = "%" + escapeIdent(name);
+        if self.namedStructTypes.hasKey(structName) {
+            panic err:illegalArgument("This context already has a struct type by that name");
```
Comment (jclark) [inline compiler/modules/print.llvm/llvm.bal]: Don't capitalize error messages

File: compiler/modules/print.llvm/common.bal (hunk 2 above)
Comment (jclark) [reply compiler/modules/print.llvm/common.bal]: I don't think it belongs: you are mapping from a name to a struct type.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/print.llvm/llvm.bal (hunk 3 above)
Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/llvm.bal]: fixed

File: compiler/modules/print.llvm/llvm.bal (hunk 4 above)
Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/llvm.bal]: fixed

**Pull request merged**